diff --git a/src/cpu/o3/iew.cc b/src/cpu/o3/iew.cc
index 21b488a..90868b7 100644
--- a/src/cpu/o3/iew.cc
+++ b/src/cpu/o3/iew.cc
@@ -143,6 +143,22 @@ IEW::regProbePoints()
 
 IEW::IEWStats::IEWStats(CPU *cpu)
     : statistics::Group(cpu, "iew"),
+    ADD_STAT(forwards_taken, statistics::units::Count::get(),
+             "number of forward branches taken"),
+    ADD_STAT(forwards_not_taken, statistics::units::Count::get(),
+             "number of forward branches not taken"),
+    ADD_STAT(total_backwards_taken, statistics::units::Count::get(),
+             "number of backward branches taken"),
+    ADD_STAT(total_backwards_not_taken, statistics::units::Count::get(),
+             "total number of backward branches not taken"),
+    ADD_STAT(mispredicted_forward_taken, statistics::units::Count::get(),
+             "number of mispredicted forwards taken"),
+    ADD_STAT(mispredicted_forward_not_taken, statistics::units::Count::get(),
+             "number of mispredicted forwards not taken"),
+    ADD_STAT(mispredicted_backward_taken, statistics::units::Count::get(),
+             "number of mispredicted backwards taken"),
+    ADD_STAT(mispredicted_backward_not_taken, statistics::units::Count::get(),
+             "number of mispredicted backwards not taken"),
     ADD_STAT(idleCycles, statistics::units::Cycle::get(),
              "Number of cycles IEW is idle"),
     ADD_STAT(squashCycles, statistics::units::Cycle::get(),
@@ -189,6 +205,7 @@ IEW::IEWStats::IEWStats(CPU *cpu)
     ADD_STAT(wbFanout, statistics::units::Rate<
                 statistics::units::Count, statistics::units::Count>::get(),
              "Average fanout of values written-back")
+
 {
     instsToCommit
         .init(cpu->numThreads)
@@ -1545,6 +1562,32 @@ IEW::updateExeInstStats(const DynInstPtr& inst)
     //
     if (inst->isControl()) {
         cpu->executeStats[tid]->numBranches++;
+	
+		Addr target_address = inst->branchTarget(tc);
+        Addr inst_address = inst -> pcState();
+			
+		if(inst->mispredicted()){
+			if(inst->readPredTaken()){
+				if(target_address > inst_address){mispredicted_forwards_taken++;}				
+				else{mispredicted_backwards_taken++;}
+			}else{
+				if(target_address > inst_address){mispredicted_forwards_not_taken++;}
+				else{mispredicted_backwards_not_taken++;}
+			}
+		}
+
+		if(inst->readPredTaken()){
+        	if(target_address > inst_address){total_forwards_taken++;}
+            else{total_backwards_taken++;}
+        }else{
+        	if(target_address > inst_address){total_forwards_not_taken++;}
+            else{total_backwards_not_taken++;}
+    	}
+
+
+
+
+
     }
 
     //
diff --git a/src/cpu/o3/iew.hh b/src/cpu/o3/iew.hh
index 1307e0c..e9f5aeb 100644
--- a/src/cpu/o3/iew.hh
+++ b/src/cpu/o3/iew.hh
@@ -422,6 +422,15 @@ class IEW
     {
         IEWStats(CPU *cpu);
 
+        statistics::Scalar forwards_taken;
+        statistics::Scalar forwards_not_taken;
+        statistics::Scalar total_backwards_taken;
+        statistics::Scalar total_backwards_not_taken;
+        statistics::Scalar mispredicted_forward_taken;
+        statistics::Scalar mispredicted_forward_not_taken;
+        statistics::Scalar mispredicted_backward_taken;
+        statistics::Scalar mispredicted_backward_not_taken;
+
         /** Stat for total number of idle cycles. */
         statistics::Scalar idleCycles;
         /** Stat for total number of squashing cycles. */
