diff --git a/.gem5-resources/riscv-spec-mcf-bin b/.gem5-resources/riscv-spec-mcf-bin
new file mode 100755
index 0000000..1efd399
Binary files /dev/null and b/.gem5-resources/riscv-spec-mcf-bin differ
diff --git a/diff.patch b/diff.patch
new file mode 100644
index 0000000..e69de29
diff --git a/src/cpu/o3/iew.cc b/src/cpu/o3/iew.cc
index 21b488a..4fd9919 100644
--- a/src/cpu/o3/iew.cc
+++ b/src/cpu/o3/iew.cc
@@ -143,6 +143,22 @@ IEW::regProbePoints()
 
 IEW::IEWStats::IEWStats(CPU *cpu)
     : statistics::Group(cpu, "iew"),
+    ADD_STAT(forwards_taken, statistics::units::Count::get(),
+             "number of forward branches taken"),
+    ADD_STAT(forwards_not_taken, statistics::units::Count::get(),
+             "number of forward branches not taken"),
+    ADD_STAT(total_backwards_taken, statistics::units::Count::get(),
+             "number of backward branches taken"),
+    ADD_STAT(total_backwards_not_taken, statistics::units::Count::get(),
+             "total number of backward branches not taken"),
+    ADD_STAT(mispredicted_forward_taken, statistics::units::Count::get(),
+             "number of mispredicted forwards taken"),
+    ADD_STAT(mispredicted_forward_not_taken, statistics::units::Count::get(),
+             "number of mispredicted forwards not taken"),
+    ADD_STAT(mispredicted_backward_taken, statistics::units::Count::get(),
+             "number of mispredicted backwards taken"),
+    ADD_STAT(mispredicted_backward_not_taken, statistics::units::Count::get(),
+             "number of mispredicted backwards not taken"),
     ADD_STAT(idleCycles, statistics::units::Cycle::get(),
              "Number of cycles IEW is idle"),
     ADD_STAT(squashCycles, statistics::units::Cycle::get(),
@@ -189,6 +205,7 @@ IEW::IEWStats::IEWStats(CPU *cpu)
     ADD_STAT(wbFanout, statistics::units::Rate<
                 statistics::units::Count, statistics::units::Count>::get(),
              "Average fanout of values written-back")
+
 {
     instsToCommit
         .init(cpu->numThreads)
diff --git a/src/cpu/o3/iew.hh b/src/cpu/o3/iew.hh
index 1307e0c..e9f5aeb 100644
--- a/src/cpu/o3/iew.hh
+++ b/src/cpu/o3/iew.hh
@@ -422,6 +422,15 @@ class IEW
     {
         IEWStats(CPU *cpu);
 
+        statistics::Scalar forwards_taken;
+        statistics::Scalar forwards_not_taken;
+        statistics::Scalar total_backwards_taken;
+        statistics::Scalar total_backwards_not_taken;
+        statistics::Scalar mispredicted_forward_taken;
+        statistics::Scalar mispredicted_forward_not_taken;
+        statistics::Scalar mispredicted_backward_taken;
+        statistics::Scalar mispredicted_backward_not_taken;
+
         /** Stat for total number of idle cycles. */
         statistics::Scalar idleCycles;
         /** Stat for total number of squashing cycles. */
